Document how we respond to security vulnerabilities and issue rewards (#529)
sourcegraph/about: docs/security.md

All development laptops have encrypted hard drives.

## How to report a security vulnerability

> If you think that you have found a security issue, please email us at <a href="mailto:security@sourcegraph.com">security@sourcegraph.com</a>. We will acknowledge receiving your report within 1 business day and will strive to send you regular updates on our progress until the issue is resolved. You may request an update by replying to the existing email thread.

Author: ```suggestion
If you think that you have found a security issue, please email us at <a href="mailto:security@sourcegraph.com">security@sourcegraph.com</a>. We will reply to high quality reports within 1 business day to acknowledge that we received them and will strive to send you regular updates on our progress until the issue is resolved. You may request an update by replying to the existing email thread. We will read, but may not respond to or issue bounties for low quality or spammy reports (e.g. those produced by automated tooling).
```

> If you think that you have found a security issue, please email us at <a href="mailto:security@sourcegraph.com">security@sourcegraph.com</a>. We will reply to high quality reports within 1 business day to acknowledge that we received them and will strive to send you regular updates on our progress until the issue is resolved. You may request an update by replying to the existing email thread. We will read, but may not respond to or issue bounties for low quality or spammy reports (e.g. those produced by automated tooling).

Reviewer: ```suggestion
If you think that you have found a security issue, please email us at <a href="mailto:security@sourcegraph.com">security@sourcegraph.com</a>. We will reply to all reports within 1 business day to acknowledge that we received them and will strive to send you regular updates on our progress until the issue is resolved. You may request an update by replying to the existing email thread. We will read, but may not respond to or issue bounties for low quality or spammy reports (e.g. those produced by automated tooling).
```

I think this wording is better because it implies that our low-quality report handling is the exception.

Author: "All" seems like a direct contradiction, so I will just drop "high quality"

```suggestion
If you think that you have found a security issue, please email us at <a href="mailto:security@sourcegraph.com">security@sourcegraph.com</a>. We will reply to reports within 1 business day to acknowledge that we received them and will strive to send you regular updates on our progress until the issue is resolved. You may request an update by replying to the existing email thread. We will read, but may not respond to low quality or spammy reports (e.g. those produced by automated tooling).
```

We provide monetary rewards, up to $10,000 USD, for security vulnerability reports. The actual reward amount is determined based on the number of customers impacted, the difficulty of exploiting the vulnerability, and the severity of the consequences (e.g. service disruption, data leakage, reputational damage to Sourcegraph) of a successful exploit.

We will send payment to a valid PayPal account after the issue is confirmed fixed or 90 days from the original report, whichever happens first. We will ask you for the name and country associated with your PayPal account.

We will not issue a reward if any of the following apply:

1. You engage in disruptive behavior on sourcegraph.com itself (e.g. spamming our system with requests, fake accounts, denial of service). Sourcegraph is [open source software](https://github.com/sourcegraph/sourcegraph), so you can [install a copy yourself](https://docs.sourcegraph.com/#quickstart-guide) and test against that instead.
1. You publically disclose a vulnerability before we have fixed it.
1. You spam us with duplicate and/or low quality vulnerability reports (e.g. copy/pasting generic issues from automatic scanning tools).

> 1. You are currently a teammate at Sourcegraph (e.g. employee, contractor, intern).

Reviewer: You may wish to add a family/friend clause here as well?

Author: It is a grey area. I added a bullet point here but softened the language at the top: f2e5ee0b7be32bbbfb6e5b1aa1fe885344e47258

> We define multipliers for each of the factors for the reward amount and apply these multipliers against our maximum bounty. For example, a vulnerability that impacts a small number of customers (0.5x), is difficult to exploit (0.5x), and would cause a lot of harm (1x), might be awarded $2,500 USD.

Reviewer 2: Is the max bounty 2,500 or 10,000?

Author: $10,000, as documented above. Would this suggested edit have clarified it for you?

```suggestion
We define multipliers for each of the factors for the reward amount and apply these multipliers against our maximum bounty ($10,000 USD). For example, a vulnerability that impacts a small number of customers (0.5x), is difficult to exploit (0.5x), and would cause a lot of harm (1x), might be awarded $2,500 USD (i.e. `$10,000 * 0.5 * 0.5 * 1.0 = $2,500`).
```

> These are only guidelines and the final award amount for any issue is at the discression of Sourcegraph.

Reviewer 2: ```suggestion
These are only guidelines and the final award amount for any issue is at the discretion of Sourcegraph.
```

| Vulnerable customers* | Reward multiplier |
|----------------------|-------------------|
| A majority (> 50%) of customers | 1x |
| A small number of customers | 0.5x |
| No customers | 0.01x |

> \* A **customer** is defined as a paid user of Sourcegraph. Please note that there are currently no paying customers on sourcegraph.com because sourcegraph.com only hosts public code and is free to use. As such, an issue that impacts all **users** on sourcegraph.com doesn't necessarily impact any **customers**.

Reviewer: This indicates that someone reporting a major vulnerability in Sourcegraph.com that leaks, say, GitHub OAuth tokens, would receive $10,000 * 0.01x or just $100, which I assume can't be correct?

Author: Yeah, multipliers are hard. I removed them.

| Difficulty of exploit | Reward multiplier |
|-----------------------|-------------------|
| Anyone can feasibly exploit this vulnerability. | 1x |
| This vulerability is difficult to exploit for one or more reasons (e.g. requires uncommon circumstances that aren't under the attacker's control). | 0.5x |

| Severity | Reward multiplier |
|----------|-------------------|

> | A successful exploit could result in material harm to Sourcegraph and/or our customers. | 1x |

Reviewer 2: The multiples strike me as false precision, since the definition of "material harm" is fairly vague and abstract. If we just want to merge this because it improves on what we have today, I'm fine with that, but perhaps better would be to replace the multiples with some examples of bugs and what bounties they would receive. I.e., provide data points.

Author: Are you against the multipliers in general or just this "harm" based factor? The goal of documenting this is to make it less vague and abstract than it is today (and to direct researchers to researching vulnerabilities in dimensions that are most valuable to us).

The multiples here are precise to the extent that they document how I think about issuing rewards when the time comes. I think it is useful to have guidelines written down so that (1) if anyone disagrees with my valuations they can be challenged and we can discuss updating the guidelines, and (2) we can delegate this more across the team.

If you or anyone else wants to own this process, I am happy to defer on methodology for how we determine rewards, but I do think we need to have **a** methodology and it needs to be documented.

> perhaps better would be to replace the multiples with some examples of bugs and what bounties they would receive. I.e., provide data points.

This isn't what other companies do, and we don't have enough examples to make this helpful in my opinion. GitLab does "customer based" maximum bounties: https://hackerone.com/gitlab (which is basically the first table of multiples).

Author: I removed the multipliers
